costmap_cspace/src/costmap_3d.cpp

```diff
@@ -184,10 +167,113 @@ class Costmap3DOFNode
         "map", 1,
         boost::bind(&Costmap3DOFNode::cbMap, this, _1, root_layer));
 
-    auto layer = costmap_->addLayer<costmap_cspace::Costmap3dLayerFootprint>(overlay_mode);
-    sub_map_overlay_ = nh_.subscribe<nav_msgs::OccupancyGrid>(
-        "map_overlay", 1,
-        boost::bind(&Costmap3DOFNode::cbMapOverlay, this, _1, layer));
+    if (nhp_.hasParam("layers"))
+    {
+      XmlRpc::XmlRpcValue layers_xml;
+      nhp_.getParam("layers", layers_xml);
+
+      if (layers_xml.getType() != XmlRpc::XmlRpcValue::TypeStruct || layers_xml.size() < 1)
+      {
+        ROS_FATAL("layers parameter must contain at least one layer config.");
+        throw std::runtime_error("layers parameter must contain at least one layer config.");
+      }
+      for (auto &layer_xml : layers_xml)
+      {
+        ROS_INFO("New layer: %s", layer_xml.first.c_str());
+        costmap_cspace::Costmap3dLayerFootprint::map_overlay_mode overlay_mode;
+        if (layer_xml.second["overlay_mode"].getType() == XmlRpc::XmlRpcValue::TypeString)
+        {
+          std::string overlay_mode_str(layer_xml.second["overlay_mode"]);
+          if (overlay_mode_str.compare("overwrite") == 0)
+            overlay_mode = costmap_cspace::Costmap3dLayerFootprint::map_overlay_mode::OVERWRITE;
+          else if (overlay_mode_str.compare("max") == 0)
+            overlay_mode = costmap_cspace::Costmap3dLayerFootprint::map_overlay_mode::MAX;
+          else
+          {
+            ROS_FATAL("Unknown overlay_mode \"%s\"", overlay_mode_str.c_str());
+            throw std::runtime_error("Unknown overlay_mode.");
+          }
+        }
+        else
+        {
+          ROS_WARN("overlay_mode of the layer is not specified. Using MAX mode.");
+          overlay_mode = costmap_cspace::Costmap3dLayerFootprint::map_overlay_mode::MAX;
+        }
+
+        std::string type;
+        if (layer_xml.second["type"].getType() == XmlRpc::XmlRpcValue::TypeString)
+        {
+          type = std::string(layer_xml.second["type"]);
+        }
+        else
+        {
+          ROS_FATAL("Layer type is not specified.");
+          throw std::runtime_error("Layer type is not specified.");
+        }
+
+        costmap_cspace::Costmap3dLayerBase::Ptr layer;
+
+        if (type == "Costmap3dLayerFootprint")
```

Comment:
Add getName to Costmap3dLayer classes and make class loader. (or use ROS pluginlib)